### ISSUE TITLE
AppCenter Distribute issue with Realme UI

### ISSUE THREAD

**Issue by MicaelMor** [microsoft/appcenter-sdk-dotnet#1418]
EDITED:

Please see comment https://github.com/microsoft/appcenter-sdk-dotnet/issues/1418#issuecomment-658261668 for further information relevant to the issue.

### **Description**

~When using the AppCenter.Distribute (version 3.3.0 and older) with the in-app update on a Realme phone with Realme UI (tested on version RMX1931EX_11_C.28 of a realme x2 pro) the app does not try to update, for the purposes of this issue the app was installed through the app center QR code download link, and further tests were conducted using the debug mode (with Distribute.SetEnabledForDebuggableBuild(true) so it would try to update while debugging), the app was at no point installed from a store.~

When using the AppCenter.Distribute (version 3.3.0 and older) with the in-app update on a Realme phone with Realme UI (tested on version RMX1931EX_11_C.28 of a realme x2 pro) the app does not try to update, for the purpose of this issue the app was installed through visual studio debugger, with debug mode update enabled (with Distribute.SetEnabledForDebuggableBuild(true) so it would try to update while debugging), further tests were conducted by installing the app through different applications, the app was at no point installed from a store.

After checking the Logcat with verbose set on, I have determined that this issue occurs due to the fact that "this.PackageManager.GetInstallerPackageName(this.PackageName)" returns "pc" when it should return null instead.

this leads to the function "InstallerUtils.isInstalledFromAppStore" returning true, when it should return false (assuming there is in fact no store whose package names returns pc"), this in turn leads for the update workflow to prematurely complete, as can be seen in the following piece of code which triggers the exit:

`           /* Don't go any further if the app was installed from an app store. */
            if (InstallerUtils.isInstalledFromAppStore(LOG_TAG, mContext)) {
                MobileCenterLog.info(LOG_TAG, "Not checking in app updates as installed from a store.");
                mWorkflowCompleted = true;
                return;
            }`

This bug does not occur in either the emulator or a xiaomi mi a2 lite in which this same app was tested on.

### **Repro Steps**

Please list the steps used to reproduce your issue.

1. Have a Realme UI phone, preferably a Realme X2 Pro (only realme device I tested on), and debug an application with AppCenter in-app update turned on.

OR

1. Use the "PackageManager.SetInstallerPackageName" function to set the InstallerPackageName to the string "pc", before calling "AppCenter.Start", this should allow the bug to be reproduced (untested).

EDITED:

Since the issue occurs every time the InstallerPackageName is not null and doesn't match any of the values found in the "LOCAL_STORES", any value here would do, Realme UI in particular seems to set InstallerPackageName with the app id from which the .apk file was installed from instead of a fixed value.

### **Details**

1. What is your app platform (Xamarin.Android or Xamarin.iOS or UWP)?
    Xamarin.Android (using F# with it through F# Fabulous)
2. If using Xamarin.Forms or if using portable/shared code to call our SDK APIs, are you using shared project, PCL code or .NET standard code for the application? Which .NET standard version or which PCL profile?
    Using Xamarin.Forms, and calling a .Net Standard 2.0 F# library, but this error is not related to that, since it happens before any calls get made to that library.
4. Which OS version did you experience the issue on?
    Android 11
5. What device version did you see this error on?  Were you using an emulator or a physical device?
    Realme X2 pro physical device.
6. What third party libraries are you using? 
    APPCenter, distribute, analytics, crashes, all 3.3.0, Fabulous 0.54.0 and its required dependencies and so on, 
7. Please enable verbose logging for your app using `AppCenter.LogLevel = LogLevel.Verbose` before your call to `AppCenter.Start(...)` and include the logs here:
  

> 07-13   23:17:02.704 | realme RMX1931 | Info | 4607 | AppCenterDistribute | Not checking in app updates as installed from a store.
-- | -- | -- | -- | -- | --
07-13 23:17:02.704 | realme RMX1931 | Debug | 4607 | AppCenterDistribute | InstallerPackageName=pc
07-13 23:17:02.704 | realme RMX1931 | Info | 4607 | Choreographer | Skipped 8 frames!  The   application may be doing too much work on its main thread.
07-13 23:17:02.704 | realme RMX1931 | Info | 4607 | AppCenterDistribute | Not checking in app updates as installed from a store.
07-13 23:17:02.704 | realme RMX1931 | Debug | 4607 | AppCenterDistribute | InstallerPackageName=pc
07-13 23:17:02.704 | realme RMX1931 | Info | 4607 | Choreographer | Skipped 8 frames!  The   application may be doing too much work on its main thread.
07-13 23:17:02.704 | realme RMX1931 | Info | 4607 | AppCenterDistribute | Not checking in app updates as installed from a store.
07-13 23:17:02.704 | realme RMX1931 | Debug | 4607 | AppCenterDistribute | InstallerPackageName=pc
07-13 23:17:02.702 | realme RMX1931 | Debug | 4607 | AppCenterDistribute | Resume distribute workflow...
07-13 23:17:02.702 | realme RMX1931 | Debug | 4607 | AppCenterDistribute | Resume distribute workflow...
07-13 23:17:02.702 | realme RMX1931 | Debug | 4607 | AppCenterDistribute | Resume distribute workflow...

### **Known Workarounds**

~Before calling "AppCenter.Start" one can check to see if "PackageManager.GetInstallerPackageName" returns the string "pc" and then use PackageManager.SetInstallerPackageName to set it to null, doing so solves the issue, since now the package name will return as null meaning it was not installed from a store.~

EDITED: Due to cert signing issues this seems to only work in debug mode, not in release mode.

### **Additional Notes**
~It is possible that "pc" might be a valid store location, but since I am unsure about that I have decided to create this issue anyway, I would like to also say that since I have a known workaround that is just a couple of lines of code this is also very much not an urgent issue on my end at all.~

~EDITED: Given that this also seems to happen with the "com.android.chrome" InstallerPackageName, it is likely due to some outdated packages I might have in my project, will have to further test this.~

Thank you.

**Comment by MicaelMor** [comment]
Just checked the issue with a brand new Xamarin project, and a brand new appcenter project, with everything updated in visual studio and nuget, and I am still facing the same issue where I am getting "Not checking in app updates as installed from a store.", despite it clearly not having been installed from a store, log bellow:

> 

Time | Device Name | Type | PID | Tag | Message
-- | -- | -- | -- | -- | --
07-14 12:58:28.572 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_analytics) pendingLogCount=1   batchTimeInterval=3000
07-14 12:58:28.572 | realme RMX1931 | Debug | 31461 | AppCenter | enqueue(group_analytics) pendingLogCount=1
07-14 12:58:28.572 | realme RMX1931 | Debug | 31461 | AppCenter | Stored a log to the Persistence database for log type   startSession with databaseId=4
07-14 12:58:28.567 | realme RMX1931 | Debug | 31461 | AppCenter | Storing a log to the Persistence database for log type   startSession with flags=1
07-14 12:58:28.567 | realme RMX1931 | Debug | 31461 | AppCenterAnalytics | onActivityResumed
07-14 12:58:28.567 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_core) pendingLogCount=1   batchTimeInterval=3000
07-14 12:58:28.567 | realme RMX1931 | Debug | 31461 | AppCenter | enqueue(group_core) pendingLogCount=1
07-14 12:58:28.567 | realme RMX1931 | Debug | 31461 | AppCenter | Stored a log to the Persistence database for log type   startService with databaseId=3
07-14 12:58:28.562 | realme RMX1931 | Info | 31461 | AppCenterDistribute | Not checking in app updates as installed from a store.
07-14 12:58:28.562 | realme RMX1931 | Debug | 31461 | AppCenterDistribute | InstallerPackageName=com.android.chrome
07-14 12:58:28.562 | realme RMX1931 | Debug | 31461 | AppCenterDistribute | Resume distribute workflow...
07-14 12:58:28.560 | realme RMX1931 | Debug | 31461 | AppCenter | Storing a log to the Persistence database for log type   startService with flags=1
07-14 12:58:28.557 | realme RMX1931 | Info | 31461 | AppCenter | Crashes service started from application.
07-14 12:58:28.557 | realme RMX1931 | Debug | 31461 | AppCenterCrashes | No previous minidump sub-folders.
07-14 12:58:28.556 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(groupErrors/one) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.556 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(groupErrors/one)
07-14 12:58:28.556 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(groupErrors) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.556 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(groupErrors)
07-14 12:58:28.556 | realme RMX1931 | Debug | 31461 | AppCenter | removeGroup(groupErrors/one)
07-14 12:58:28.556 | realme RMX1931 | Debug | 31461 | AppCenter | removeGroup(groupErrors)
07-14 12:58:28.556 | realme RMX1931 | Info | 31461 | AppCenter | Analytics service started from application.
07-14 12:58:28.555 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_analytics_critical/one)   pendingLogCount=0 batchTimeInterval=3000
07-14 12:58:28.555 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_analytics_critical/one)
07-14 12:58:28.555 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_analytics_critical)   pendingLogCount=0 batchTimeInterval=3000
07-14 12:58:28.555 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_analytics_critical)
07-14 12:58:28.555 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_analytics/one) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.554 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_analytics/one)
07-14 12:58:28.554 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_analytics) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.554 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_analytics)
07-14 12:58:28.554 | realme RMX1931 | Debug | 31461 | AppCenter | removeGroup(group_analytics/one)
07-14 12:58:28.554 | realme RMX1931 | Debug | 31461 | AppCenter | removeGroup(group_analytics)
07-14 12:58:28.554 | realme RMX1931 | Info | 31461 | AppCenter | Distribute service started from application.
07-14 12:58:28.554 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_distribute/one) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.553 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_distribute/one)
07-14 12:58:28.553 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_distribute) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.553 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_distribute)
07-14 12:58:28.553 | realme RMX1931 | Debug | 31461 | AppCenter | removeGroup(group_distribute/one)
07-14 12:58:28.553 | realme RMX1931 | Debug | 31461 | AppCenter | removeGroup(group_distribute)
07-14 12:58:28.553 | realme RMX1931 | Debug | 31461 | AppCenter | App Center initialized.
07-14 12:58:28.552 | realme RMX1931 | Debug | 31461 | AppCenter | checkPendingLogs(group_core) pendingLogCount=0   batchTimeInterval=3000
07-14 12:58:28.552 | realme RMX1931 | Debug | 31461 | AppCenter | addGroup(group_core)
07-14 12:58:28.552 | realme RMX1931 | Info | 31461 | AppCenter | Changed maximum database size to 10485760 bytes.
07-14 12:58:28.550 | realme RMX1931 | Debug | 31461 | AppCenter | Network has been connected.
07-14 12:58:28.550 | realme RMX1931 | Debug | 31461 | AppCenter | Network 100 is available.
07-14 12:58:28.548 | realme RMX1931 | Debug | 31461 | AppCenter | Loaded stored sessions:   {1594727848596=1594727848596//1594727848596,   1594727848628=1594727848628/4c6448bf-22c8-42e0-944b-b4b90006b19f/1594727848596}
07-14 12:58:28.548 | realme RMX1931 | Debug | 31461 | AppCenter | Cannot read instrumentation variables in a non-test   environment.
07-14 12:58:28.547 | realme RMX1931 | Debug | 31461 | AppCenter | Cannot read instrumentation variables in a non-test   environment.
07-14 12:58:28.547 | realme RMX1931 | Info | 31461 | AppCenter | App Center SDK configured successfully.
07-14 12:58:28.541 | realme RMX1931 | Info | 31461 | AppCenterCrashes | Set up Xamarin crash handlers.
07-14 12:58:28.528 | realme RMX1931 | Debug | 31461 | AppCenterXamarin | No named identifier found in appSecret; using as-is

**Comment by olmisa** [comment]
Hello @MicaelMor 
Thanks for getting in touch!
I have created PR with a fix. I have added a new store location.

**Comment by MicaelMor** [comment]
> Hello @MicaelMor
> Thanks for getting in touch!
> I have created PR with a fix. I have added a new store location.
> Thanks

Hi, thanks for the quick reply.

By mistake, I didn't properly update the original post after investing this issue more thoroughly, so I left out an important piece of information, which means the PR with the fix won't solve the problem completely, sorry about that.

This doesn't just happen with "pc", it seems like Realme UI gives InstallerPackageName a different value for each location the app is installed from, being that the value is the app package name.
So while if the app is copied from the PC using say Visual Studio it gives the name "pc", if one installs it from android chrome app https://play.google.com/store/apps/details?id=com.android.chrome&hl=en it gives that app package name value which for chrome is "com.android.chrome".
Giving another example if the same file were to be installed from google file app https://play.google.com/store/apps/details?id=com.google.android.apps.nbu.files&hl=en it gives the package name of that app which is "com.google.android.apps.nbu.files".

In short the issue with appcenter in-app updates on the Realme UI OS cannot be solved by adding variables to the LOCAL_STORES, since the value changes based on the app that was used to run the .apk file.

Once again apologies for not making this known sooner, but I thought I had properly updated the original post.

**Comment by MicaelMor** [comment]
I have some thoughts regarding this problem that I will leave here:

It seems to me like this issue is primarily caused by incorrect behaviour from Realme UI which changes InstallerPackageName based on install source, something that does not seem to occur with other phones I have tested with, which keep the value null when not installed from stores, hence why the code works on those phones.

There seems to be a couple of ways to fix this:

The first is from the appcenter user side, where the programmer making the app using app center could manually check the InstallerPackageName name and change it as needed, I have tried to do this, and while it works in debug, in a full release it fails due to the apk signing issues, I am not sure if this is a hard limit making this a non solution, or if it is just simply me doing something wrong.

The second way to fix this to work with Realme UI is to change from the appcenter sdk side, the most obvious way for me (as someone that is unfamiliar with the code base), is to create a list of actual digital stores that the user could set.
When that list is not set (null or empty depending on desired behaviour here), the appcenter update works as it does now, if however the list is set, the InstallerPackageName is compared to the elements of that list, if a match is found then the app update quits and doesn't update (since this would be a list of actual digital stores), if however no match is found then the update proceeds.

**Comment by olmisa** [comment]
@MicaelMor could you, please, provide a full list of the _InstallerPackageName_ values which could be produced by Realme UI (I have no such device)? We are discussing what could be done with this issue.
Thanks.

**Comment by MicaelMor** [comment]
Hi,

For my own use case the following list of names would solve the issue:

"com.android.chrome"
"org.mozilla.firefox"
"com.google.android.apps.nbu.files"
"pc"

Since these would cover the 4 means of installation I would see myself use in the near future for my apps (google chrome, firefox, files by google and a pc).

However, I must stress that this list might not solve the issue for everyone that might use a Realme UI powered device with AppCenter Distribute, since Realme UI device could produce any value for InstallerPackageName.

For example if tomorrow someone released an app to open files that had the PackageName "com.somefileapp.android" and that app was used to open the .apk of an app that uses AppCenter Distribute, this .apk would now be returning "com.somefileapp.android" as the InstallerPackageName.

Essentially modifying the list of LOCAL_STORES would be patching the issue (and in turn fixing it for me), but it would not be solving it completely for everyone, since other users might for example download and install the .apk of their apps using Microsoft Edge, which would return an InstallerPackageName of "com.microsoft.emmx"

Thank you.

**Comment by olmisa** [comment]
Hello @MicaelMor ,
we decided to update LOCAL_STORES. If it will be not enough for some cases then we return to this issue.
Thanks

**Comment by AnastasiaKubova** [comment]
The fix has been released in [3.4.0](https://github.com/microsoft/appcenter-sdk-dotnet/releases/tag/3.4.0).
